### ISSUE TITLE
Fuzzuf crashes with floating-point exception while fuzzing with aflfast/afl

### ISSUE THREAD

**Issue by key-moon** [fuzzuf/fuzzuf#24]
## Relation to existing issues
N/A

## Relation to existing roadmaps
N/A

## Bug explanation

Fuzzuf crashes with floating-point exception while fuzzing with `aflfast` or `afl`. I tested only these fuzzer, but maybe this is not the specific problem for these fuzzers.

## Steps to reproduce

1. Execute fuzzuf with the `aflfast` fuzzer
    -  `fuzzuf aflfast --in_dir=seeds --out_dir=out -- ./put @@`
    - If you want to, I will append the put that I tested.
2. Wait some times (10 min or so)

### Environment

- fuzzuf version or commit ID: 0412d2
- OS: Ubuntu 20.04.3 LTS on Windows 10 x86_64
- Kernel: 5.10.43.3-microsoft-standard-WSL2

## Expected behaviors
No error occurs.

## Error messages, stack traces, or logs

For your information, 浮動小数点例外 means the "floating point exception".

```sh
$ fuzzuf aflfast --in_dir=seeds --out_dir=out -- ./put @@
[*] Scanning 'seeds'...
[*] Creating hard links for all input files...
[*] Attempting dry run with 'id:000000,orig:seed.md'...
    len = 103, map size = 581, exec speed = 4441 us
[+] All test cases processed.

                         fuzzuf american fuzzy lop (fast) 2.57b (put)

┌─ process timing ─────────────────────────────────────┬─ overall results ─────┐
│        run time : 0 days, 0 hrs, 11 min, 1 sec       │  cycles done : infty  │
│   last new path : 0 days, 0 hrs, 0 min, 45 sec       │  total paths : 1086   │
│ last uniq crash : none seen yet                      │ uniq crashes : 0      │
│  last uniq hang : none seen yet                      │   uniq hangs : 0      │
├─ cycle progress ────────────────────┬─ map coverage ─┴───────────────────────┤
│  now processing : 122.0 (11.23%)    │    map density : 0.93% / 5.52%         │
│ paths timed out : 0 (0.00%)         │ count coverage : 2.74 bits/tuple       │
├─ stage progress ────────────────────┼─ findings in depth ────────────────────┤
│  now trying : auto extras (over)    │ favored paths : 272 (25.05%)           │
│ stage execs : 84/99 (84.85%)        │  new edges on : 519 (47.79%)           │
│ total execs : 108k                  │ total crashes : 0 (0 unique)           │
│  exec speed : 169.5/sec             │  total tmouts : 0 (0)                  │
├─ fuzzing strategy yields ───────────┴───────────────┬─ path geometry ────────┤
│   bit flips : 85/3296, 37/3292, 16/3284             │    levels : 3          │
│  byte flips : 0/412, 0/408, 3/400                   │   pending : 1083       │
│ arithmetics : 79/23.0k, 0/1051, 0/0                 │  pend fav : 272        │
│  known ints : 10/2123, 1/11.3k, 0/17.6k             │ own finds : 1085       │
│  dictionary : 0/0, 0/0, 0/291                       │  imported : n/a        │
│       havoc : 854/32.8k, 0/0                        │ stability : 100.00%    │
│        trim : 0.00%/148, 0.00%                      ├────────────────────────┘
浮動小数点例外────────────────────────────────────────┘          [cpu000: 22%]
$ 
```

<details><summary>detailed gdb trace information</summary>

```c
Program received signal SIGFPE, Arithmetic exception.
0x00007f467c544075 in fuzzuf::algorithm::afl::util::UR (limit=0, rand_fd=8) at /home/keymoon/ricsec/fuzzuf/algorithms/afl/afl_util.cpp:50
50          return random() % limit;
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
───────────────────────────────────────[ REGISTERS ]────────────────────────────────────────
*RAX  0x5e966345
*RBX  0x2
*RCX  0x0
*RDX  0x0
*RDI  0x7f467b8e0740 (unsafe_state) —▸ 0x7f467b8e0234 (randtbl+116) ◂— 0x8f1ca3625a60b629
*RSI  0x7ffca3249954 ◂— 0x2ae469005e966345
 R8   0x0
*R9   0x7f467b8e0240 (pa_next_type) ◂— 0x8
*R10  0x55f96b93e010 ◂— 0x7000300020004
*R11  0x7f467b8e0be0 (main_arena+96) —▸ 0x55f96bc84f30 ◂— 0x0
*R12  0xb
*R13  0x67
 R14  0x0
*R15  0x0
*RBP  0x7ffca32499a0 —▸ 0x7ffca32499c0 —▸ 0x7ffca3249a00 —▸ 0x7ffca3249ae0 —▸ 0x7ffca3249b60 ◂— ...
*RSP  0x7ffca3249970 ◂— 0x3fe0000000000000
*RIP  0x7f467c544075 ◂— idiv   rcx
─────────────────────────────────────────[ DISASM ]─────────────────────────────────────────
 ► 0x7f467c544075    idiv   rcx
    ↓
   0x7f467c544075    idiv   rcx
─────────────────────────────────────[ SOURCE (CODE) ]──────────────────────────────────────
In file: /mnt/c/Users/keymoon/ricsec/fuzzuf/algorithms/afl/afl_util.cpp
   45
   46         using option::AFLTag;
   47         using option::GetReseedRng;
   48         rand_cnt = (GetReseedRng<AFLTag>() / 2) + (seed[1] % GetReseedRng<AFLTag>());
   49     }
 ► 50     return random() % limit;
   51 }
   52
   53 /* Describe all the integers with five characters or less */
   54
   55 std::string DescribeInteger(u64 val) {
─────────────────────────────────────────[ STACK ]──────────────────────────────────────────
00:0000│ rsp 0x7ffca3249970 ◂— 0x3fe0000000000000
01:0008│     0x7ffca3249978 ◂— 0x8
02:0010│     0x7ffca3249980 ◂— 0x3ff0000000000000
03:0018│     0x7ffca3249988 ◂— 0x16789bf72ae46900
04:0020│     0x7ffca3249990 ◂— 0x2
05:0028│     0x7ffca3249998 ◂— 0xb /* '\x0b' */
06:0030│ rbp 0x7ffca32499a0 —▸ 0x7ffca32499c0 —▸ 0x7ffca3249a00 —▸ 0x7ffca3249ae0 —▸ 0x7ffca3249b60 ◂— ...
07:0038│     0x7ffca32499a8 —▸ 0x7f467cc8ba72 ◂— leave
```

```c
#0  0x00007f467c544075 in fuzzuf::algorithm::afl::util::UR (limit=0, rand_fd=8) at /home/keymoon/ricsec/fuzzuf/algorithms/afl/afl_util.cpp:50
#1  0x00007f467cc8ba72 in fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::ChooseBlockLen(unsigned int)::{lambda(unsigned int)#1}::operator()(unsigned int) const (this=0x7ffca3249d70, limit=0) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_mutator.hpp:58
#2  0x00007f467cc8bafe in fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::ChooseBlockLen (this=0x7ffca3249d70, limit=102) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_mutator.hpp:61
#3  0x00007f467cc8877f in Mutator<fuzzuf::algorithm::aflfast::option::AFLFastTag>::Havoc<unsigned int (*)(std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&, std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&), fuzzuf::algorithm::afl::routine::mutation::HavocBaseTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::DoHavoc(fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, int, int)::{lambda(int, unsigned char*&, unsigned int&)#1}>(unsigned int, std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&, std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&, unsigned int (*)(std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&, std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&), fuzzuf::algorithm::afl::routine::mutation::HavocBaseTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::DoHavoc(fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, int, int)::{lambda(int, unsigned char*&, unsigned int&)#1}) (this=0x7ffca3249d70, stacking=16, extras=std::vector of length 0, capacity 0, a_extras=std::vector of length 1, capacity 1 = {...}, case_distrib=0x7f467c544ee8 <fuzzuf::algorithm::afl::util::HavocCaseDistrib(std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&, std::vector<fuzzuf::algorithm::afl::dictionary::AFLDictData, std::allocator<fuzzuf::algorithm::afl::dictionary::AFLDictData> > const&)>, custom_cases=...) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/mutator/mutator.hpp:501
#4  0x00007f467cc838a9 in fuzzuf::algorithm::afl::routine::mutation::HavocBaseTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::DoHavoc (this=0x55f96b956a10, mutator=..., stage_name="havoc", stage_short="havoc", perf_score=3200, stage_max_multiplier=1024, stage_idx=15) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp:415
#5  0x00007f467cc7e420 in fuzzuf::algorithm::afl::routine::mutation::HavocTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::operator() (this=0x55f96b956a10, mutator=...) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp:460
#6  0x00007f467cc809fe in HierarFlowNodeImpl<bool (fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&), bool (unsigned char const*, unsigned int)>::operator()(fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&) (this=0x55f96b956a60, args#0=...) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_node_impl.hpp:104
#7  0x00007f467cc86755 in HierarFlowRoutine<bool (std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>), bool (fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&)>::runAllChildren(fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&) (this=0x55f96b9561e0, args#0=...) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_routine.hpp:122
#8  0x00007f467cc80a96 in HierarFlowRoutine<bool (std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>), bool (fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&)>::CallSuccessors(fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&) (this=0x55f96b9561e0, args#0=...) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_routine.hpp:101
#9  0x00007f467cc80b42 in fuzzuf::algorithm::afl::routine::other::ApplyRandMutsTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::operator() (this=0x55f96b9561e0, testcase=std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase> (use count 3, weak count 0) = {...}) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_other_hierarflow_routines.hpp:504
#10 0x00007f467cc826b0 in HierarFlowNodeImpl<bool (std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>), bool (fuzzuf::algorithm::afl::AFLMutatorTemplate<fuzzuf::algorithm::aflfast::AFLFastState>&)>::operator()(std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>) (this=0x55f96b956240, args#0=std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase> (empty) = {...}) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_node_impl.hpp:104
#11 0x00007f467cc875cb in HierarFlowRoutine<void (), bool (std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>)>::runAllChildren(std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>) (this=0x55f96b953c70, args#0=std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase> (use count 3, weak count 0) = {...}) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_routine.hpp:122
#12 0x00007f467cc82bc3 in HierarFlowRoutine<void (), bool (std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>)>::CallSuccessors(std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>) (this=0x55f96b953c70, args#0=std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase> (empty) = {...}) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_routine.hpp:101
#13 0x00007f467cc82e6e in fuzzuf::algorithm::afl::routine::other::SelectSeedTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::operator() (this=0x55f96b953c70) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_other_hierarflow_routines.hpp:113
#14 0x00007f467cc82f8d in HierarFlowNodeImpl<void (), bool (std::shared_ptr<fuzzuf::algorithm::aflfast::AFLFastTestcase>)>::operator()() (this=0x55f96b955c70) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_node_impl.hpp:104
#15 0x00007f467cc49aee in HierarFlowRoutine<void (), void ()>::runAllChildren() (this=0x55f96b952e20) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_routine.hpp:122
#16 0x00007f467cc40fb4 in HierarFlowRoutine<void (), void ()>::CallSuccessors() (this=0x55f96b952e20) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_routine.hpp:101
#17 0x00007f467cc44e6b in fuzzuf::hierarflow::ProxyRoutine<void ()>::operator()() (this=0x55f96b952e20) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/utility.hpp:86
#18 0x00007f467cc45063 in HierarFlowNodeImpl<void (), void ()>::operator()() (this=0x55f96b952e80) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_node_impl.hpp:104#19 0x00007f467cc4a9c9 in HierarFlowNode<void (), void (), true>::operator()<void>() (this=0x55f96b952a30) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/hierarflow/hierarflow_node.hpp:277
#20 0x00007f467cc83092 in fuzzuf::algorithm::afl::AFLFuzzerTemplate<fuzzuf::algorithm::aflfast::AFLFastState>::OneLoop (this=0x55f96b952a20) at /home/keymoon/ricsec/fuzzuf/include/fuzzuf/algorithms/afl/templates/afl_fuzzer.hpp:125
#21 0x000055f96a6323b1 in main (argc=7, argv=0x7ffca324b9c8) at /home/keymoon/ricsec/fuzzuf/cli/cli_entry_point.cpp:72
#22 0x00007f467b71c0b3 in __libc_start_main (main=0x55f96a631d10 <main(int, char const**)>, argc=7, argv=0x7ffca324b9c8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffca324b9b8) at ../csu/libc-start.c:308
#23 0x000055f96a631ace in _start ()
```
</details>

## Supplementary information (optional)

With debugging with gdb, I found that this is caused in afl_util.cpp:50 by the limit being 0.

https://github.com/fuzzuf/fuzzuf/blob/0412d2013f343b761edd76115670012c6e0002c0/algorithms/afl/afl_util.cpp#L39-L51

This is called from mutator.hpp:58, which is also called from mutator.hpp:61. This indicates that `rlim` is also being 0 when this crash happens, which is most likely caused by the fact that `state.queue_cycle` is also being 0.

https://github.com/fuzzuf/fuzzuf/blob/0412d2013f343b761edd76115670012c6e0002c0/include/fuzzuf/algorithms/afl/templates/afl_mutator.hpp#L50-L65

### Update: 02/18 20:20
The reason why this crash only occurs after 10 minutes is L54. I recommend deleting this line when you investigate this crash.

https://github.com/fuzzuf/fuzzuf/blob/0412d2013f343b761edd76115670012c6e0002c0/include/fuzzuf/algorithms/afl/templates/afl_mutator.hpp#L52-L56

https://github.com/fuzzuf/fuzzuf/blob/838ef5df6b7b20d202717946342bf6b4f3908297/include/fuzzuf/algorithms/afl/afl_state.hpp#L215-L217

---
> Authors of this issue must fill the checklist below when they submit it

#### Mandatory entries
- [x] The title of this issue clearly summarizes the bug
- [x] The relations to existing issues and roadmaps are stated if they exist
- [x] Each entry above has been filled

---
> A repository maintainer will set up an asignee for each issue.
> Asignee must check this issue from the perspective of the checklist below before going through a full-fledged bug-fix, and leave the initial comment.
> Copy to your comment and fill out the checklist on that occasion.

#### Optional entries
- [ ] If the bug is going to be fixed, it has been commented
- [ ] If it is not going to be fixed and this issue is closed, it has been commented

#### Mandatory entries
- [ ] Have checked relationships to existing issues and roadmaps
- [ ] The minimum required information to reproduce the bug and spot the root cause is included

---
> If an assignee is going to solve the task for this issue, consider the policy and divide the procedures into the checklist.
> Once divided, write down the checklist to the comment.
> Each step for the procedures must correspond to one single pull request. Once the PR is made, update the checklist.
> If the procedures are remained undecided, or need additional reviews, update the checklist appropriately with the discussion here.

#### Procedure checklist
- [ ] Write down the step 1
    - If you have choices rejected, add as an item here with the reasons of rejection
- [ ] Write down the step 2
    - Ditto
- [ ] ...

---
> If an assignee finishes solving the task for this issue, copy the following checklist to the comment and fill it out.

#### Optional entries
- [ ] If you have remaining tasks, relating issues have been created or updated
- [ ] If the changes against this issue affect the milestone (e.g. TODO.md), the necessity of version upgrading has already been discussed
- [ ] If new tests are required to guarantee the changes against this issue, they have been split to the checklist
- [ ] If entries in the checklist are hard to understand, that has been pointed out
- [ ] If the discussion concludes that the design concept should be documented appropriately, an assignee has checked additional documents and source codes
- [ ] If there were uncertainties during the discussion, all of them have been resolved

#### Mandatory entries
- [ ] The procedure checklist is not required to update any longer
- [ ] The PRs correspond to each entries of the checklist have adopted the best way to solve the problems as long as an assignee can come up with
- [ ] All the PRs correspond to each entries of the checklist have already been merged
- [ ] This issue can be closed

--
> If an assignee agrees to close the issue without solving the tasks relating it, describe the reasons in the comment.
> Then, copy the following checklist to the comment and fill them out.

#### Mandatory entries
- [ ] The reason to close the issue has been agreed among assignees
- [ ] This issue can be closed

**Comment by key-moon** [comment]
According to commit https://github.com/fuzzuf/fuzzuf/commit/b37c7d058737db8f0dcdfc87fb477050bb3e9dad, `queue_cycle` being zero is not an unintended behavior. So we should clamp the value instead of taking the min. 
Edit: This commit actually fixed this issue.

**Comment by key-moon** [comment]
The fix for this issue is included in #17.

**Comment by potetisensei** [comment]
I close this issue because https://github.com/fuzzuf/fuzzuf/pull/17 is merged.
